### ISSUE TITLE
Tips on training

### ISSUE THREAD

**Issue by vyouman** [espnet/espnet#669]
I did experiments on wsj0_mix dataset but got worse results with the current recipe. I tried to change the default parameters (different initialization seeds and different scheduled sampling probs) but didn't improve so much. The WER of my current best model is 41.46. It's far from the result 25.40 in the published paper. Can you give me some tips on training the CTC/Attention model? Which parameters may have a great influence? Thanks.

> /home/user/vyouman/espnet/egs/wsj_mix/asr1/../../../utils/compute-perm-free-error.sh --num_spkrs 2 --wrd true exp/tr_pytorch_vggblstmp_sde1_rece2_subsample1_2_2_1_1_unit1024_proj1024_d1_unit300_location_spatrue_aconvc10_aconvf100_mtlalpha0.2_adadelta_sampprob0.0_bs10_mli800_mlo150_seed8_lsmunigram0.05_delta/decode_tt_beam30_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_1layer_unit1000_sgd_bs300_word65000/result_r1h1.wrd.txt exp/tr_pytorch_vggblstmp_sde1_rece2_subsample1_2_2_1_1_unit1024_proj1024_d1_unit300_location_spatrue_aconvc10_aconvf100_mtlalpha0.2_adadelta_sampprob0.0_bs10_mli800_mlo150_seed8_lsmunigram0.05_delta/decode_tt_beam30_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_1layer_unit1000_sgd_bs300_word65000/result_r1h2.wrd.txt exp/tr_pytorch_vggblstmp_sde1_rece2_subsample1_2_2_1_1_unit1024_proj1024_d1_unit300_location_spatrue_aconvc10_aconvf100_mtlalpha0.2_adadelta_sampprob0.0_bs10_mli800_mlo150_seed8_lsmunigram0.05_delta/decode_tt_beam30_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_1layer_unit1000_sgd_bs300_word65000/result_r2h1.wrd.txt exp/tr_pytorch_vggblstmp_sde1_rece2_subsample1_2_2_1_1_unit1024_proj1024_d1_unit300_location_spatrue_aconvc10_aconvf100_mtlalpha0.2_adadelta_sampprob0.0_bs10_mli800_mlo150_seed8_lsmunigram0.05_delta/decode_tt_beam30_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_1layer_unit1000_sgd_bs300_word65000/result_r2h2.wrd.txt
> Total Scores: (#C #S #D #I) 63687 29226 5700 5963
> Error Rate:   41.46 
> Total Utts:  3000

> /home/user/vyouman/espnet/egs/wsj_mix/asr1/../../../utils/compute-perm-free-error.sh --num_spkrs 2 exp/tr_pytorch_vggblstmp_sde1_rece2_subsample1_2_2_1_1_unit1024_proj1024_d1_unit300_location_spatrue_aconvc10_aconvf100_mtlalpha0.2_adadelta_sampprob0.0_bs10_mli800_mlo150_seed8_lsmunigram0.05_delta/decode_tt_beam30_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_1layer_unit1000_sgd_bs300_word65000/result_r1h1.txt exp/tr_pytorch_vggblstmp_sde1_rece2_subsample1_2_2_1_1_unit1024_proj1024_d1_unit300_location_spatrue_aconvc10_aconvf100_mtlalpha0.2_adadelta_sampprob0.0_bs10_mli800_mlo150_seed8_lsmunigram0.05_delta/decode_tt_beam30_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_1layer_unit1000_sgd_bs300_word65000/result_r1h2.txt exp/tr_pytorch_vggblstmp_sde1_rece2_subsample1_2_2_1_1_unit1024_proj1024_d1_unit300_location_spatrue_aconvc10_aconvf100_mtlalpha0.2_adadelta_sampprob0.0_bs10_mli800_mlo150_seed8_lsmunigram0.05_delta/decode_tt_beam30_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_1layer_unit1000_sgd_bs300_word65000/result_r2h1.txt exp/tr_pytorch_vggblstmp_sde1_rece2_subsample1_2_2_1_1_unit1024_proj1024_d1_unit300_location_spatrue_aconvc10_aconvf100_mtlalpha0.2_adadelta_sampprob0.0_bs10_mli800_mlo150_seed8_lsmunigram0.05_delta/decode_tt_beam30_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_1layer_unit1000_sgd_bs300_word65000/result_r2h2.txt
> Total Scores: (#C #S #D #I) 475101 71003 52183 40836
> Error Rate:   27.42 
> Total Utts:  3000 

> 
![Screenshot from 2019-04-02 10-46-36](https://user-images.githubusercontent.com/9965927/55372537-a5cf5700-5534-11e9-984b-24df4a26a719.png)
![Screenshot from 2019-04-02 10-47-27](https://user-images.githubusercontent.com/9965927/55372563-be3f7180-5534-11e9-97b9-d0ddfd1eaddf.png)
The accuracy of the validation set was about 88%.

Verbose of the training log at the last epoch 13.
![Screenshot from 2019-04-02 15-47-46](https://user-images.githubusercontent.com/9965927/55385097-ef349c00-555e-11e9-9925-02cabd733b9d.png)
Some decoding results
![Screenshot from 2019-04-02 16-11-00](https://user-images.githubusercontent.com/9965927/55386430-088b1780-5562-11e9-9ee2-08825a7ddcdd.png)

**Comment by sw005320** [comment]
> The accuracy of the validation set was about 88%.

@simpleoier, what is your validation accuracy? As far as I remember, you got more than 90%. The training seems to be wrong then.

**Comment by simpleoier** [comment]
Yes, I got nearly 90% (89.8% to be precise).

**Comment by sw005320** [comment]
OK, the difference is not so large than I expected. It may be decoding. @vyouman, what kind of LM are you using? Same with the recipe one?

**Comment by simpleoier** [comment]
Actually, we have discussed the LM thing before. The decoding results without the LM are also different between mine and hers. Her result is about 10% worse than mine, if I recall it correctly.

**Comment by vyouman** [comment]
@sw005320 The 1000-unit word-level RNNLM as the recipe. The training corpus for the language model is training transcription of train_si284 and external LM training corpus provided by wsj1. The wer of the model without LM on the test set is about 51%.

**Comment by sw005320** [comment]
OK. I'll run the recipe to check it. It may take time.

**Comment by vyouman** [comment]
@sw005320 OK, I'll wait for your result. Thanks!

**Comment by vyouman** [comment]
@sw005320 Hi, have you got the result? I've tried to tune other parameters (encoding units, encoding layers, lsm_weight, penalty and etc) last week, but it didn't make much difference.

**Comment by sw005320** [comment]
Not yet. The cluster was busy due to Interspeech deadline. I think I'll get the result in a day.

**Comment by sw005320** [comment]
I got the result, which is a similar range to @vyouman's ones

```
write CER (or TER) results in exp/tr_pytorch_vggblstmp_sde1_rece2_subsample1_2_2_1_1_unit1024_proj1024_d1_unit300_location_spafalse_aconvc10_aconvf100_mtlalpha0.2_adadelta_sampprob0.0_bs10_mli800_mlo150_lsmunigram0.05_delta/decode_tt_beam30_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_1layer_unit1000_sgd_bs300_word65000/result_r*h*.txt
Total Scores: (#C #S #D #I) 474887 72827 50582 40295
Error Rate:   27.36
Total Utts:  3000
write WER results in exp/tr_pytorch_vggblstmp_sde1_rece2_subsample1_2_2_1_1_unit1024_proj1024_d1_unit300_location_spafalse_aconvc10_aconvf100_mtlalpha0.2_adadelta_sampprob0.0_bs10_mli800_mlo150_lsmunigram0.05_delta/decode_tt_beam30_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_1layer_unit1000_sgd_bs300_word65000/result_r*h*.wrd.txt
Total Scores: (#C #S #D #I) 64335 28844 5434 5739
Error Rate:   40.58
Total Utts:  3000
Finished
```
The validation accuracy is `validation/main/acc": 0.8765130770576979`, and slightly lower than the on @simpleoier reported. This might be a bug in the training. I'm also concerning about the scoring part. @simpleoier, we have to fix this issue.

**Comment by vyouman** [comment]
@simpleoier @sw005320 I used the Espnet 0.2v wsj0_2mix recipe @simpleoier gave me to run the multi-speaker recipe. Yeah, it seemed like the new version (0.3v) of Espnet degraded the performance of multi-talker ASR. With Espnet 0.2v, I got the result like this.
> WER
> Total Scores: (#C #S #D #I) 75160 18802 4651 3588
> Error Rate:   27.42
> Total Utts:  3000

> CER
> Total Scores: (#C #S #D #I) 515117 43949 39221 22507
> Error Rate:   17.66
> Total Utts:  3000 

Interestingly, I compared the performance of wsj recipe (single-speaker ASR) between Espnet 0.2 and Espnet 0.3 under the same configuration, the wer and cer of the new version are slightly better than the old version (about 0.2~0.3).

**Comment by sw005320** [comment]
Thanks. Which pytorch version are you using? We observed the significant CTC loss difference between old and new versions.

**Comment by vyouman** [comment]
@sw005320 I'm using pytorch 0.4.

**Comment by sw005320** [comment]
for both Espnet 0.2 and Espnet 0.3?

**Comment by vyouman** [comment]
@sw005320 Yeah. I used the warp-ctc loss.
